[Merged by Bors] - feat(algebra/lie/basic): define derived length and semisimple Lie algebras

ocfnash opened the pull request:
We also provide proofs of some basic characterisations

---

If this is merged then I would like to follow up with a (much smaller) PR doing something similar for reductive Lie algebras and then finally refactor along the lines outlined [here](https://github.com/leanprover-community/mathlib/pull/5794#discussion_r560050825).

ocfnash committed:
feat(algebra/lie/basic): define derived length and semisimple Lie algebras

We also provide proofs of some basic characterisations

jcommelin commented on src/algebra/lie/basic.lean:
```diff
@@ -474,6 +480,33 @@ lie_module.is_trivial.trivial x m
 abbreviation is_lie_abelian (L : Type v) [has_bracket L L] [has_zero L] : Prop :=
 lie_module.is_trivial L L
 
+lemma lie_abelian_of_injection {R : Type u} {L₁ : Type v} {L₂ : Type w}
```
We have similar things for `monoid`, `comm_ring`, etc, and there the declarations are put in the `function.{injective|surjective}` namespace.
So `function.injective.is_lie_abelian`, and then you can use it as `h1.is_lie_abelian h2`.

ocfnash replied:
Not even out for review and you're helping!

Thanks, done :-)

ocfnash committed:
Minor updates

ocfnash committed:
CR suggestion, better namespacing: `function.injective.is_lie_abelian`, `function.surjective.is_lie_abelian`

ocfnash commented on src/algebra/lie/basic.lean:
```diff
@@ -474,6 +480,33 @@ lie_module.is_trivial.trivial x m
+begin
+  split; intros h,
+  { exact e.symm.injective.is_lie_abelian h, },
+  { exact @function.injective.is_lie_abelian R L₁ L₂ _ _ _ _ _ e e.injective h, },
```
I don't understand why the following doesn't work instead:
```lean
exact e.injective.is_lie_abelian h,
```
Bizarrely, the following ridiculous code works though:
```lean
exact e.symm.symm.injective.is_lie_abelian h,
```
I will investigate if I have time this evening.

ocfnash replied:
I just pushed [a commit](https://github.com/leanprover-community/mathlib/pull/5930/commits/9e881ed336a5e7c05e1426c4cc3e4495b684e5bd) which I claim resolves this issue.

ocfnash commented on src/algebra/lie/basic.lean:
```diff
@@ -1743,6 +1969,45 @@ begin
+
+/-- A semisimple Lie algebra is one with trivial radical. -/
+class is_semisimple : Prop :=
+(semisimple : radical R L = ⊥)
```
It's worth noting that no less a person than [Jim Humphreys claimed](https://mathoverflow.net/questions/149391/on-radicals-of-a-lie-algebra#comment383669_149391) there is no agreement on standard definition of semisimple outside of characteristic zero.

Nevertheless I am a least consistent with Seligman who literally wrote the book on modular Lie algebras. He says:
![image](https://user-images.githubusercontent.com/7734364/106145617-aacfb280-616d-11eb-89a2-c63a7dfd2b5e.png)

jcommelin replied:
Feel free to write a short remark about this in the docstring. I'm sure I will forget otherwise.

ocfnash replied:
Done!

ocfnash committed:
Improve docstring for `is_semisimple`.

Also adding a reference and other minor docstring tweaks.

jcommelin commented on src/order/conditionally_complete_lattice.lean:
```diff
@@ -483,6 +483,29 @@ end
+  rw nat.Inf_eq_zero, right, assumption,
+end
+
+lemma Inf_interval_above_eq_succ_iff {s : set ℕ}
```
This is the only lemma where I'm not sure if this is the mathlibbest way to do things. The rest of the PR looks good to me.
@PatrickMassot @urkud can you take a look at this please?

ocfnash replied:
Thanks, for the review and agreed: this lemma is where I felt least sure about what was best.

PatrickMassot replied:
I just had a look at how I would prove this lemma and I wrote the following. I'm not sure it's better, but at least it extracts two more lemmas that don't seem to be there already. That `nat.has_Inf` is pretty specific, using both that `nat` is bounded below and the fact it's a discrete order, I'm not sure we should spend too much time generalizing this.
```lean
lemma nat.eq_Ici_of_nonempty_of_upward_closed {s : set ℕ} (hs : s.nonempty)
  (hs' : ∀ (k₁ k₂ : ℕ), k₁ ≤ k₂ → k₁ ∈ s → k₂ ∈ s) : s = Ici (Inf s) :=
ext (λ n, ⟨λ H, nat.Inf_le H, λ H, hs' (Inf s) n H (nat.Inf_mem hs)⟩)

lemma nat.nonempty_of_Inf_eq_succ {s : set ℕ} {k : ℕ} (h : Inf s = k + 1) : s.nonempty :=
nat.nonempty_of_pos_Inf (h.symm ▸ (nat.succ_pos k) : Inf s > 0)

lemma Inf_interval_above_eq_succ_iff {s : set ℕ}
  (hs : ∀ (k₁ k₂ : ℕ), k₁ ≤ k₂ → k₁ ∈ s → k₂ ∈ s) (k : ℕ) :
  Inf s = k + 1 ↔ k + 1 ∈ s ∧ k ∉ s :=
begin
  split,
  { intro H,
    rw show s = Ici (Inf s),
      from nat.eq_Ici_of_nonempty_of_upward_closed (nat.nonempty_of_Inf_eq_succ H) hs,
    simp [H] },
  { rintro ⟨H, H'⟩,
    rw [nat.Inf_def (⟨_, H⟩ : s.nonempty), nat.find_eq_iff],
    exact ⟨H, λ n hnk hns, H' $ hs n k (nat.lt_succ_iff.mp hnk) hns⟩ } 
end
```

ocfnash replied:
Thank you very much @PatrickMassot. I have just pushed (a slightly-modified version of) your code above.

I also agree that the two new lemmas are worth having in their own right.

jcommelin replied:
@PatrickMassot thanks a lot!

jcommelin reviewed (approved):
LGTM, apart from one lemma, where I would like an opinion from another maintainer.

ocfnash committed:
Improve proof of `Inf_interval_above_eq_succ_iff`

Co-authored-by: Patrick Massot <patrickmassot@free.fr>

ocfnash committed:
Rename `Inf_interval_above_eq_succ_iff` --> `Inf_upward_closed_eq_succ_iff`

For consistency with `eq_Ici_of_nonempty_of_upward_closed`

ocfnash committed:
Force coercion via a morhpism in statement of `lie_algebra.equiv.injective`

Similarly for `surjective` and `bijective`.

I only partly understand but this seems to make the job of the elaborator
easier when using these lemmas as propositions about morphisms (rather than
equivalences).

ocfnash committed:
Vanity golfing

jcommelin reviewed (approved):
Thanks :tada:

bors merge
